File: godot/Scenes/V2Graph/Graph.cs
using System.Linq;
using Godot;

namespace GCCC.Scenes.V2Graph;

/// <summary>
/// Represents a graph visualization node in the scene tree
/// </summary>
public partial class Graph : Node
{
    [Export]
    private int _Resolution = 100;

    private float Step => 2.0f / _Resolution;

    private Vector3 Size => Vector3.One * Step;

    private readonly StandardMaterial3D _material = new()
    {
        AlbedoColor = Colors.Red
    };

    // Called when the node enters the scene tree for the first time.
    public override void _Ready()
    {
        for (int i = 0; i < _Resolution * _Resolution; i++)
        {
            AddChild(new CsgBox3D
            {
                Size = Size,
                Material = _material
            });
        }
    }

    // Called every frame. 'delta' is the elapsed time since the previous frame.
    public override void _Process(double delta)
    {
        var time = Time.GetTicksMsec() / 1000.0f;

        for (var i = 0; i < GetChildCount(); i++)
        {
            if (GetChild(i) is not CsgBox3D point) continue;

            var (u, v) = (i % _Resolution, i / _Resolution);
            var (x, z) = (
                u * 2.0f / _Resolution - 1.0f,
                v * 2.0f / _Resolution - 1.0f 
            );

            point.Position = new(
                x,
                2.0f * Mathf.Sin(time + x + z),
                z
            );
        }
    }
}

File: rust/src/graph.rs
use godot::classes::{CsgBox3D, Node3D, StandardMaterial3D, Time};
use godot::prelude::*;

#[derive(GodotClass)]
#[class(init, base=Node3D)]
struct Graph {
    #[export]
    #[init(val = 200)]
    resolution: i32,

    #[base]
    base: Base<Node3D>,
}

#[godot_api]
impl INode3D for Graph {
    fn ready(&mut self) {
        let size = self.size();
        let material = Graph::material();

        for _ in 0..(self.resolution * self.resolution) {
            let mut node = CsgBox3D::new_alloc();
            node.set_size(size);
            node.set_material(&material);
            self.base_mut().add_child(&node);
        }
    }

    fn process(&mut self, _delta: f64) {
        let time = Time::singleton().get_ticks_msec() as f32 / 1000.0;

        for i in 0..self.base().get_child_count() {
            let mut node = match self
                .base()
                .get_child(i)
                .and_then(|n| n.try_cast::<CsgBox3D>().ok())
            {
                Some(node) => node,
                None => continue,
            };

            let (u, v) = (i % self.resolution, i / self.resolution);
            let (x, z) = (u as f32 * self.step() - 1.0, v as f32 * self.step() - 1.0);
            node.set_position(Vector3::new(x, 2.0 * (time + x + z).sin(), z));
        }
    }
}

impl Graph {
    fn step(&self) -> f32 {
        2.0 / self.resolution as f32
    }

    fn size(&self) -> Vector3 {
        Vector3::ONE * self.step()
    }

    fn material() -> Gd<StandardMaterial3D> {
        let mut material = StandardMaterial3D::new_gd();
        material.set_albedo(Color::GREEN);
        material
    }
}
